src/include/serial.i

```diff
@@ -237,7 +237,7 @@ __wt_update_serial(WT_SESSION_IMPL *session, WT_PAGE *page, WT_UPDATE **srch_upd
      * Check if our update is still permitted.
      */
     while (!__wt_atomic_cas_ptr(srch_upd, upd->next, upd)) {
-        if ((ret = __wt_txn_update_check(session, upd->next = *srch_upd)) != 0) {
+        if ((ret = __wt_txn_update_check(session, cbt, upd->next = *srch_upd)) != 0) {
```

Comment:
This seems to be causing some problems - should we have `__wt_update_serial` return `EBUSY` if the compare and swap doesn't succeed, and allow the caller to figure out how to setup for a retry?

There are three callers:
* Rollback to stable, which I think is using update_serial primarily to make sure the memory footprint and dirty status are set correctly. It should never get an EBUSY.
* __wt_row_modify, which has quite a complex set of conditions in figuring out visibility checking, already involving a call to txn_update_check.
* __wt_col_modify, similar but different to row_modify.

It looks like col_modify can already be neatly retryed, and row_modify would need a small amount of code shuffling to make that work.

Reply:
It is not a problem here to recheck when cas fail. The error case is actually that cas succeed but it shouldn't. Imagine a case there are two sessions that want to delete the same key concurrently and there is no update on the update chain for the key. Both sessions pass the conflict check and are good to proceed. The first session successfully put a tombstone on the update chain. Then the page is evicted and now the key is reconciled with a stop time pair on the page and no update on the update chain. Finally, the second session tries to proceed by appending a tombstone to the update chain, which should lead to a conflict. But the cas here will be successful and thus causing the failure we see.

Reply:
> Then the page is evicted

Can a page be evicted while a session is trying to put an update on an update chain? It should hold a hazard pointer which would block the page from being evicted.

Reply:
The updates are done in several steps:
Read the current value
Check conflicts
Modify the key

Between each step, the page can be evicted.

Reply:
> Between each step, the page can be evicted.

Let's talk - I don't think that should be possible.